File: Elixir/faultInjector.ex
defmodule FaultInjector do
  def start(mode, node) do
    Process.register(spawn(FaultInjector, :injector, [mode, [], node]), :injector)
  end
  def stop() do
    Process.exit(Process.whereis(:injector), :finish)
  end
  def injector(mode, proc_list, node) do
    receive do
      {"Start", sup_list, num_fails, injection_interval} ->
        IO.puts("Generating faultload: #{num_fails} per second, Mode: #{mode}")

        num_sups = length(sup_list)
        max_kills = 5
        num_killers = div(num_fails, max_kills)
        leftover = rem(num_fails, max_kills)
        sup_per_killer = if num_sups>num_killers do
          remainder = if rem(num_sups, num_killers) != 0 do
            1
          else
            0
          end
          div(num_sups, num_killers)+remainder
        else
          1
        end
        killer_list = case mode do
          :burst ->
            spawnkiller(node, node, max_kills, sup_list, div(5000, num_killers), mode, num_killers, sup_per_killer)
          :stairs ->
            Process.sleep(6*injection_interval)
            killer_list = for n <- :lists.seq(1, num_sups, sup_per_killer),  do: spawn(__MODULE__, :stairs_killer, [node, max_kills, :lists.sublist(sup_list, n, sup_per_killer), injection_interval])
            IO.puts("#{num_sups}: #{length(killer_list)}")
            killer_list
          :uniform ->
            spawnkiller(node, node, max_kills, sup_list, div(div(1000, max_kills), num_killers), mode, num_killers, sup_per_killer)
          :rand ->
            spawnkiller(node, node, max_kills, sup_list, :rand.uniform(max(div(div(1000, max_kills), num_killers), 1)), mode, num_killers, sup_per_killer)
          _ ->
            []
        end
        injector(mode, [spawn(__MODULE__, :killer, [node, leftover, sup_list, mode]) | killer_list], node)

      "Stop" ->
        for pid <- proc_list, do: Process.exit(pid, :normal)
        injector(mode, [], node)
      end
  end
  def spawnkiller(_, _, _, _, _, _, 0, _) do
    []
  end
  def spawnkiller(node, node, num_fails, sup_list, interval, mode, num, step) do
    pid = spawn(__MODULE__, :killer, [node, num_fails, sup_list, mode])
    Process.sleep(interval)
    {head, tail} = :lists.split(step, sup_list)
    [pid | spawnkiller(node, node, num_fails, :lists.append(tail, head), interval, mode, num-1, step)]
  end

  def killer(node, num_fails, sup_list, :burst) do
    interval = 5
    start_time = :erlang.system_time(:millisecond)
    new_sup_list = burst_kill(node, sup_list, num_fails*interval)
    sleep = (interval * 1000) - (:erlang.system_time(:millisecond)-start_time)
    if sleep>0 do
      Process.sleep(sleep)
    end
    killer(node, num_fails, new_sup_list, :burst)
  end


  def killer(node, num_fails, [s1 | sup_list], :uniform) do
    start_time = :erlang.system_time(:millisecond)
    kill_process(node, s1)
    sleep = div(1000, num_fails) - (:erlang.system_time(:millisecond) - start_time)
    if sleep>0 do
      Process.sleep(sleep)
    end
    killer(node, num_fails, :lists.append(sup_list, [s1]), :uniform)
  end
  def killer(node, num_fails, sup_list, :rand) do
    type = :rand.uniform(2)
    size = :rand.uniform(10)

    new_sup_list = if type == 1 do
      rand_kill(node, sup_list, size, div(1000, num_fails))
    else
      start_time = :erlang.system_time(:millisecond)
      new_sup_list = burst_kill(node, sup_list, size)
      sleep = div(1000, num_fails)*size - :erlang.system_time(:millisecond) + start_time
      if sleep>0 do
        Process.sleep(sleep)
      end
      new_sup_list
    end
    killer(node, num_fails, new_sup_list, :rand)
  end

  def rand_kill(_, sup_list, 0, _) do
    sup_list
  end
  def rand_kill(node, [s1|sup_list], n, interval) do
    start_time = :erlang.system_time(:millisecond)
    kill_process(node, s1)
    elapsed_time = :erlang.system_time(:millisecond) - start_time
    if elapsed_time<interval do
      Process.sleep(interval - elapsed_time)
    end
    rand_kill(node, :lists.append(sup_list, [s1]), n-1, interval)
  end

  def stairs_killer(node, num_fails, sup_list, injection_interval) do
    Process.sleep(5*injection_interval)
    new_sup_list = burst_kill(node, sup_list, num_fails)
    killer(node, num_fails, new_sup_list, :stairs)
  end

  def burst_kill(_, sup_list, 0) do
    sup_list
  end
  def burst_kill(node, [s1 | sup_list], n) do
    kill_process(node, s1)
    burst_kill(node, :lists.append(sup_list, [s1]), n-1)
  end
  def kill_process(node, sup) do
    p_list = Supervisor.which_children(sup)
    {_, target, _, _} = :lists.nth(:rand.uniform(length(p_list)), p_list)
    status = :rpc.call(node, Process, :alive?, [target])
    if status do
      Process.exit(target, :kill)
    else
      kill_process(node, sup)
    end
  end
end

File: Erlang/faultInjector.erl
-module(faultInjector).
-compile(export_all).

start(Mode, Node) ->
    register(injector, spawn(faultInjector, injector, [Mode, [], Node])).

stop() -> 
    exit(whereis(injector), finish).

injector(Mode, ProcList, Node) ->
    receive
        {"Start", SupList, NFailures, InjectionInterval} ->
            io:format("Generating faultload: ~p per second, Mode: ~p~n", [NFailures, Mode]),

            NSups = length(SupList),                         % Number of Supervisors in the system
            MaxKills = 5,                                    % Max number of times a killer kills per second
            NKillers = NFailures div MaxKills,               % Number of killers required for current fault load
            Leftover = NFailures rem MaxKills,               % Extra killer needed if NFailures is not perfectly divisible by MaxKills
            
            if
                NSups > NKillers ->
                    if
                        NSups rem NKillers /= 0 ->
                            SupPerKiller = (NSups div NKillers) + 1;
                        true ->
                            SupPerKiller = NSups div NKillers
                    end;
                true ->
                    SupPerKiller = 1
            end,

            if 
                Mode == burst ->
                    KillerList = spawnkiller(Node, MaxKills, SupList, trunc(5000/NKillers), Mode, NKillers, SupPerKiller);
                Mode == stairs ->
                    timer:sleep(6*InjectionInterval),
                    KillerList = [spawn(?MODULE, stairsKiller, [Node, MaxKills, lists:sublist(SupList, N, SupPerKiller), InjectionInterval]) || N <- lists:seq(1, NSups, SupPerKiller)],
                    io:format("~p:~p~n", [NSups, length(KillerList)]);
                Mode == uniform ->
                    KillerList = spawnkiller(Node, MaxKills, SupList, trunc((1000/MaxKills)/NKillers), Mode, NKillers, SupPerKiller);
                Mode == rand ->
                    KillerList = spawnkiller(Node, MaxKills, SupList, rand:uniform(max(trunc((1000/MaxKills)/NKillers), 1)), Mode, NKillers, SupPerKiller);
                true ->
                    KillerList = []
            end,
            injector(Mode, [spawn(?MODULE, killer, [Node, Leftover, SupList, Mode]) | KillerList], Node);

        "Stop" ->
            [exit(Pid, normal) || Pid <- ProcList],
            injector(Mode, [], Node)
    end.

stairskiller(Node, NF, SupList, InjectionInterval) ->
    timer:sleep(5*InjectionInterval),
    NewSupList = burstKill(Node, SupList, NF),
    killer(Node, NF, NewSupList, stairs).

spawnkiller(_, _, _, _, _, 0, _) -> [];

spawnkiller(Node, NF, SupList, Interval, Mode, N, L) ->
    Pid = spawn(?MODULE, killer, [Node, NF, SupList, Mode]),
    timer:sleep(Interval),
    {HeadList, TailList} = lists:split(L, SupList),
    [Pid | spawnkiller(Node, NF, lists:append(TailList, HeadList), Interval, Mode, N-1, L)].

% Burst fault load specifications
% Single or Multiple Supervisor specification
killer(Node, NF, SupList, burst) ->
    Interval = 5,

    StartTime = erlang:system_time(millisecond),
    NewSupList = burstKill(Node, SupList, NF*Interval),
    Sleep = (Interval * 1000) - (erlang:system_time(millisecond) - StartTime),

    if
        Sleep > 0 ->
            timer:sleep(Sleep);
        true ->
            ok
    end,

    killer(Node, NF, NewSupList, burst);


% Uniform fault load specification    
killer(Node, NF, [S1 | SupList], uniform) ->
    StartTime = erlang:system_time(millisecond),
    killProcess(Node, S1),
    Sleep = (1000 / NF) - (erlang:system_time(millisecond) - StartTime),
    
    if
        Sleep > 0 ->
            timer:sleep(trunc(Sleep));
        true ->
            ok
    end,
    
    killer(Node, NF, lists:append(SupList, [S1]), uniform);

% Random fault load specification
killer(Node, NF, SupList, rand) ->
    Type = rand:uniform(2),
    Size = rand:uniform(10),

    if
        Type == 1 ->
            NewSupList = randKill(Node, SupList, Size, trunc(1000/NF));
        true ->
            StartTime = erlang:system_time(millisecond),
            NewSupList = burstKill(Node, SupList, Size),
            Sleep = ((1000 / NF) * Size) - (erlang:system_time(millisecond) - StartTime),
            if
                Sleep > 0 ->
                    timer:sleep(trunc(Sleep));
                true ->
                    ok
            end
    end,
    killer(Node, NF, NewSupList, rand).

randKill(_, SupList, 0, _) -> SupList;

randKill(Node, [S1 | SupList], N, Interval) ->
    StartTime = erlang:system_time(millisecond),
    killProcess(Node, S1),
    ElapsedTime = erlang:system_time(millisecond) - StartTime,

    if
        ElapsedTime < Interval ->
            timer:sleep(Interval - ElapsedTime);
        true ->
            ok
    end,
    randKill(Node, lists:append(SupList, [S1]), N-1, Interval).

burstKill(_, SupList, 0) -> SupList;

burstKill(Node, [S1 | SupList], N) ->
    killProcess(Node, S1),
    burstKill(Node, lists:append(SupList, [S1]), N-1).

% Process killer specifications
% Single Supervisor
killProcess(Node, Sup) ->
    PList = supervisor:which_children(Sup),
    {_, Target, _, _} = lists:nth(rand:uniform(length(PList)), PList),
    Status = rpc:call(Node, erlang, is_process_alive, [Target]),
    
    if
        Status ->
            exit(Target, kill);
        true ->
            killProcess(Node, Sup)
    end.